src/more_ops.rs

```diff
@@ -877,3 +864,45 @@ pub fn op_point_add(a: &mut Allocator, input: NodePtr, max_cost: Cost) -> Respon
     let total: G1Affine = total.into();
     new_atom_and_cost(a, cost, &total.to_compressed())
 }
+
+pub fn op_coinid(a: &mut Allocator, input: NodePtr, _max_cost: Cost) -> Response {
+    let args = Node::new(a, input);
+    check_arg_count(&args, 3, "coinid")?;
+
+    let parent_coin = atom(args.first()?, "coinid")?;
+    if parent_coin.len() != 32 {
+        return args.err("coinid: invalid parent coin id (must be 32 bytes)");
+    }
+    let args = args.rest()?;
+    let puzzle_hash = atom(args.first()?, "coinid")?;
+    if puzzle_hash.len() != 32 {
+        return args.err("coinid: invalid puzzle hash (must be 32 bytes)");
+    }
+    let args = args.rest()?;
+    let amount = atom(args.first()?, "coinid")?;
+    if !amount.is_empty() {
```

Comment:
You might consider checking the assembly generated... it's possible the optimizer might realize you don't use the return value. Also, since it likely just goes into a register anyway, it's probably not particularly expensive.

(I may be being a bit naïvely optimistic about how good the optimizers are.)

Reply:
I looked into splitting the validation and the parsing of the integer and realized that another important difference is that `coinid` requires the amount to be in canonical integer form. i.e. it does not allow any redundant leading zero. This is an important rule since it affects the hash.

The only shared check is really that the value is not negative. There are some utility functions to sanitize integers in `chia_rs`, for parsing conditions. Perhaps in the future (especially as we have more operators in here) it might make sense to move those into this repo and have some stricter conventions for validating arguments. I don't think there are enough places we do this in right now though.

Even your suggestion of calling `uint_atom::<8>()` just for the validation won't work, because redundant leading zeros would still have to be checked.